### PR TITLE
 Flytekit Auth system overhaul and pretty printing upgrade

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR introduces a new Oauth2 handling system in flytekit, that can be used in non flytekit python libraries that wish to use a standardized flow for flytekit. The standard flow is as defined in the RFC and a few additional supported methods.

The system uses client side - [grpc.Interceptors](https://grpc.github.io/grpc/python/grpc.html#client-side-interceptor). for fine control on the auth flow. It also introduces exception wrapping and retrying interceptor.

As a last time, it improves the terminal output when using pyflyte. All the output is not distilled and the error is traced correctly. Eventually we could link to docs for the errors.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

### CODE REVIEW

**Comment by kumare3** [comment]
Help wanted. 
cc @bethebunny / @kazesberger / @ByronHsu / @flixr / @honnix if some of you can test.
Especially @honnix you have a command based auth system

File: flytekit/clients/auth/auth_client.py [hunk 1]
```diff
@@ -87,12 +96,28 @@ def do_GET(self):
...
+                self.wfile.write(
+                    bytes(
+                        f"""
+                <html>
+                <head>
+                    <title>Oauth2 authentication Flow</title>
+                </head>
+                <body>
+                    <h1>Log in successful to {self.server.remote_metadata.endpoint}</h1>
+                </body></html>""",
```
Comment (EngHabu) [inline flytekit/clients/auth/auth_client.py:112]: nit: add `please close this window` or something like that..

File: flytekit/clients/auth/authenticator.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        logging.debug("Starting external process to generate id token. Command {}".format(self._cmd))
+        try:
+            output = subprocess.run(self._cmd, capture_output=True, text=True, check=True)
+        except subprocess.CalledProcessError as e:
+            logging.error("Failed to generate token from command {}".format(self._cmd))
+            raise AuthenticationError("Problems refreshing token with command: " + str(e))
+        self._creds = Credentials(output.stdout.strip())
+
+
+class ClientCredentialsAuthenticator(Authenticator):
```
Comment (EngHabu) [inline flytekit/clients/auth/authenticator.py:147]: We should also add Device flow at some point...

File: flytekit/clients/grpc_utils/wrap_exception_interceptor.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class RetryExceptionWrapperInterceptor(grpc.UnaryUnaryClientInterceptor, grpc.UnaryStreamClientInterceptor):
+    def __init__(self, max_retries: int = 3):
+        self._max_retries = 3
+
+    @staticmethod
+    def _raise_if_exc(request: typing.Any, e: Union[grpc.Call, grpc.Future]):
+        if e.code() == grpc.StatusCode.UNAUTHENTICATED:
+            raise FlyteAuthenticationException() from e
```
Comment (EngHabu) [inline flytekit/clients/grpc_utils/wrap_exception_interceptor.py:23]: Am I missing something, where is this handled?

**Review by EngHabu — DISMISSED** [review]
Thank you!

File: flytekit/clis/sdk_in_container/pyflyte.py [hunk 4]
```diff
@@ -28,7 +34,46 @@ def validate_package(ctx, param, values):
...
+        if isinstance(e, FlyteInvalidInputException):
+            click.secho("Request rejected by the API, due to Invalid input.", fg="red")
+            click.secho(f"\tReason: {str(e)}", dim=True)
+            click.secho(f"\tInput Request: {MessageToJson(e.request)}", dim=True)
+            return
+        click.secho(f"Failed with Exception: Reason: {e._ERROR_CODE}", fg="red")  # noqa
+        click.secho(f"Underlying Exception: {e.__cause__}", dim=True)
+        return
+
+    if isinstance(e, grpc._channel._InactiveRpcError):  # noqa
```
Comment (honnix) [inline flytekit/clis/sdk_in_container/pyflyte.py:52]: `grpc._channel` needs to be imported.

**Comment by honnix** [comment]
I did a quick test. I managed to run a task via `FlyteRemote` and I verified that auth token was created and propagated correctly.

File: flytekit/clients/auth/auth_client.py (hunk 1 above)
Comment (kumare3) [reply flytekit/clients/auth/auth_client.py:112]: will do, i will make it a configurable thing, also add Flyte icon

File: flytekit/clients/auth/authenticator.py (hunk 2 above)
Comment (kumare3) [reply flytekit/clients/auth/authenticator.py:147]: wdym?

File: flytekit/clients/grpc_utils/wrap_exception_interceptor.py (hunk 3 above)
Comment (kumare3) [reply flytekit/clients/grpc_utils/wrap_exception_interceptor.py:23]: this is only an exception beautifier. This is not handled anywhere, this is once all auth tries fail

File: flytekit/clients/auth/authenticator.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+class CommandAuthenticator(Authenticator):
+    """
+    This Authenticator retreives access_token using the provided command
+    """
+
+    def __init__(self, command: typing.List[str], header_key: str = None):
+        self._cmd = command
+        if not self._cmd:
+            raise ValueError("Command cannot be empty for command authenticator")
+        super().__init__(header_key)
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/clients/auth/authenticator.py:129]: thank you for the catch, updating it

File: flytekit/clients/auth/auth_client.py [hunk 6]
```diff
@@ -269,38 +288,52 @@ def request_access_token(self, auth_code):
...
+            return self._request_access_token(auth_code)
+        finally:
+            server_process.terminate()
+
+    def refresh_access_token(self, credentials: Credentials) -> Credentials:
+        if credentials.refresh_token is None:
             raise ValueError("no refresh token available with which to refresh authorization credentials")
 
         resp = _requests.post(
             url=self._token_endpoint,
```
Comment (goyalankit) [inline flytekit/clients/auth/auth_client.py:325]: We should consider passing `verify` switch to request calls as in: https://github.com/flyteorg/flytekit/pull/1509

Comment (kumare3) [reply flytekit/clients/auth/auth_client.py:314]: let me incorporate your change

**Review by wild-endeavor — APPROVED** [review]

**Comment by eapolinario** [comment]
I was able to repro the test errors in a docker container (e.g. `docker run --rm -it -v $(pwd):/root python:3.10-buster "bash"` and `make setup && make unit_test`).

Installing `keyrings.alt` makes the tests pass. That said, I'm not sure yet why we should require that package.

**Review by wild-endeavor — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]
